Fix Bad file descriptor error when using reload or workers flag and multiprocessing

likeanaxon opened the pull request:
Fix for #877. 

Note: Reopening `sys.stdin` in the child process might not even be necessary, but at any rate it works if we use `open('/dev/stdin')`. A warning is produced if for some reason `/dev/stdin` cannot be opened.

likeanaxon committed:
==[ Fix for #877 ]==

likeanaxon committed:
==[ Fixed formatting issues ]==

likeanaxon committed:
==[ Formatting, take 2 ]==

likeanaxon committed:
==[ Formatting, take 3 ]==

likeanaxon committed:
==[ Removed unused os import; Fixed empty except ]==

euri10 commented:
I tested an app with a `pdb.set_trace()` in it and can use pdb with both workers and the reload flags, however I'm not very familiar with pdb and it's use so I'd like someone else to ponder here on what we should do. 

the above fix seems reasonable to me, another pair (or more :)) of eyes would be amazing.

florimondmanca commented on uvicorn/subprocess.py:
```diff
@@ -47,12 +39,13 @@ def subprocess_started(config, target, sockets, stdin_fileno):
-        sys.stdin = os.fdopen(stdin_fileno)
+    # This is required for some debugging environments.
+    try:
+        sys.stdin = open("/dev/stdin")
```
Does /dev/stdin exist on all platforms?

likeanaxon replied:
@florimondmanca Good question. I believe that it exists on OSX, but Windows is a different beast altogether, and I think it will require a special case (based on [this SO answer](https://stackoverflow.com/a/62913474); thank you @euri10 for pointing it out in the first place). Unfortunately, I don't have access to a Windows machine to test it. My goal was to get it to work without crashing first, and then work on a cross-platform solution.

Since the file descriptor for `stdin` is `0` on *nix and [apparently also on Windows](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fileno?view=msvc-160), I can replace `open("/dev/stdin")` with `os.fdopen(0)` - would that be OK?

likeanaxon replied:
Follow-up: I went ahead and switched to `os.fdopen(0)`. Let me know if you need me to make any further changes.

likeanaxon committed:
==[ Reintroduced os.fdopen ]==
- Replaced open("/dev/stdin") with with os.fdopen(0) as the file descriptor for stdin is 0 on both *nix and Windows.

likeanaxon commented:
@florimondmanca @euri10 Is there anything else I can do to progress this?

likeanaxon committed:
Merge branch 'master' into bad_file_descriptor_fix

florimondmanca commented on uvicorn/subprocess.py:
```diff
@@ -47,12 +40,13 @@ def subprocess_started(config, target, sockets, stdin_fileno):
-        sys.stdin = os.fdopen(stdin_fileno)
+    # This is required for some debugging environments.
+    try:
+        sys.stdin = os.fdopen(0)
```
I think I'm a bit confused. Does this version fixes the issue you were seeing? Asking because I would expect stdin.fileno() to return 0, as I believe is traditional on Unix systems, so this code would do the same as initially... ?

Were you able to see the value of the file descriptor that was throwing "Bad file descriptor"?

Do we need to add `os.dup(stdin_fileno)` to create a copy rather than reusing it directly?

likeanaxon replied:
>  Does this version fixes the issue you were seeing?

Yes, it does.

> I would expect stdin.fileno() to return 0, as I believe is traditional on Unix systems

I had exactly the same expectation.

> Were you able to see the value of the file descriptor that was throwing "Bad file descriptor"?

Yes, I was. In the original code, `stdin.fileno()` returns values greater than 2 (such as 9, 10 or 11; I mentioned that in the [original report](https://github.com/encode/uvicorn/issues/877#issuecomment-742133901)). The same thing happens if I use `os.dup`.

I am not sure if the issue runs deeper than that (for example, an issue with the way `stdin.fileno()` retrieves the file descriptor for `stdin`), but the fix definitely works - I've been using my branch for a couple of weeks now.

I would be curious to see if you can reproduce the issue - there is an MRE in the [bug report](https://github.com/encode/uvicorn/issues/877#issue-759008860).

florimondmanca reviewed (commented):
Leaving a question I'm basing off this SO thread... https://stackoverflow.com/questions/32199552/what-is-sys-stdin-fileno-in-python

Kludex commented:
Closing as stale.
